master/node_selector.go

```diff
@@ -274,42 +275,54 @@ func (s *AvailableSpaceFirstNodeSelector) Select(ns *nodeSet, excludeHosts []str
 	nodes := ns.getNodes(s.nodeType)
 	sortedNodes := make([]Node, 0)
 	nodes.Range(func(key, value interface{}) bool {
-		sortedNodes = append(sortedNodes, value.(Node))
+		node := value.(Node)
+		if contains(excludeHosts, node.GetAddr()) {
+			return true
+		}
+		if !canAllocPartition(node) {
+			return true
+		}
+		sortedNodes = append(sortedNodes, node)
 		return true
 	})
+
 	// if we cannot get enough nodes, return error
 	if len(sortedNodes) < replicaNum {
-		err = fmt.Errorf("action[%vNodeSelector::Select] no enough hosts,replicaNum:%v  MatchNodeCount:%v  ",
+		err = fmt.Errorf("action[%vNodeSelector::Select] no enough writable hosts,replicaNum:%v  MatchNodeCount:%v  ",
 			s.GetName(), replicaNum, len(sortedNodes))
 		return
 	}
 	// sort nodes by available space
 	sort.Slice(sortedNodes, func(i, j int) bool {
 		return s.getNodeAvailableSpace(sortedNodes[i]) > s.getNodeAvailableSpace(sortedNodes[j])
 	})
-	nodeIndex := 0
-	// pick first N nodes
-	for i := 0; i < replicaNum && nodeIndex < len(sortedNodes); i++ {
-		selectedIndex := len(sortedNodes)
-		// loop until we get a writable node
-		for nodeIndex < len(sortedNodes) {
-			node := sortedNodes[nodeIndex]
-			nodeIndex += 1
-			if canAllocPartition(node) {
-				if excludeHosts == nil || !contains(excludeHosts, node.GetAddr()) {
-					selectedIndex = nodeIndex - 1
-					break
-				}
+
+	excludedNodes := make([]Node, 0)
+	distinctIpSet := make(map[string]struct{})
+	for len(orderHosts) < replicaNum {
+		// select set of nodes with distinct ip in each iteration, try to avoid multiple replicas of a partition locate on same machine
+		for i := 0; i < len(sortedNodes); i++ {
```

Comment:
1. Because after each inner loop， number of unselected nodes in sortedNodes changed
2. Yes， if we don't have enough nodes with distinct ip  to satisfy the replica requirement，then we will select nodes with same ip

 

Consider the below case：

Machine1 has 3 dn process (IP1:17310, IP1:18310, IP1:19310)

Machine2 has 2 dn process (IP2:17310, IP2:18310)

Machine3 has 1 dn process (IP3:17310)

1. when create 3 replicas dp, we will select IP1:17310, IP2:17310, IP3:17310 (inner loop execute once and stop)
2. when create 5 replicas dp, we will select IP1:17310, IP2:17310, IP3:17310,IP1:18310, IP2:18310
    1.  inner loop execute twice, first loop select out IP1:17310, IP2:17310, IP3:17310
    2.  because nodes with distinct ip  can’t satisfy the replica requirement, need second loop  , and select out IP1:18310, IP2:18310

Reply:
> 1. Because after each inner loop， number of unselected nodes in sortedNodes changed
> 2. Yes， if we don't have enough nodes with distinct ip  to satisfy the replica requirement，then we will select nodes with same ip
> 
> Consider the below case：
> 
> Machine1 has 3 dn process (IP1:17310, IP1:18310, IP1:19310)
> 
> Machine2 has 2 dn process (IP2:17310, IP2:18310)
> 
> Machine3 has 1 dn process (IP3:17310)
> 
> 1. when create 3 replicas dp, we will select IP1:17310, IP2:17310, IP3:17310 (inner loop execute once and stop)
> 2. when create 5 replicas dp, we will select IP1:17310, IP2:17310, IP3:17310,IP1:18310, IP2:18310
>    
>    1. inner loop execute twice, first loop select out IP1:17310, IP2:17310, IP3:17310
>    2. because nodes with distinct ip  can’t satisfy the replica requirement, need second loop  , and select out IP1:18310, IP2:18310